File: Whatsapp-Agent/src/WhatsAppAgent.js
import pkg from 'whatsapp-web.js';
const { Client, LocalAuth, MessageMedia } = pkg;
import qrcode from 'qrcode-terminal';
import fs from 'fs';
import path from 'path';

class WhatsAppAgent {
  constructor() {
    this.client = null;
    this.isReady = false;
    this.messageHandlers = [];
    
    this.initializeClient();
  }

  initializeClient() {
    // Initialize WhatsApp client with local authentication
    this.client = new Client({
      authStrategy: new LocalAuth({
        clientId: 'whatsapp-agent',
        dataPath: './.wwebjs_auth'
      }),
      puppeteer: {
        headless: true,
        args: [
          '--no-sandbox',
          '--disable-setuid-sandbox',
          '--disable-dev-shm-usage',
          '--disable-accelerated-2d-canvas',
          '--no-first-run',
          '--no-zygote',
          '--single-process',
          '--disable-gpu'
        ]
      },
      webVersionCache: {
        type: 'remote',
        remotePath: 'https://raw.githubusercontent.com/wppconnect-team/wa-version/main/html/2.2412.54.html',
      }
    });

    this.setupEventHandlers();
  }

  setupEventHandlers() {
    // QR Code generation
    this.client.on('qr', (qr) => {
      console.log('\n================================');
      console.log('📱 SCAN THIS QR CODE WITH WHATSAPP');
      console.log('================================\n');
      qrcode.generate(qr, { small: true });
      console.log('\n================================');
      console.log('Open WhatsApp on your phone');
      console.log('Go to Settings > Linked Devices');
      console.log('Tap "Link a Device" and scan the QR code above');
      console.log('================================\n');
    });

    // Authentication success
    this.client.on('authenticated', () => {
      console.log('✅ Authentication successful!');
    });

    // Authentication failure
    this.client.on('auth_failure', (msg) => {
      console.error('❌ Authentication failed:', msg);
    });

    // Client is ready
    this.client.on('ready', async () => {
      console.log('✅ WhatsApp Agent is ready!');
      this.isReady = true;
      
      const info = this.client.info;
      console.log(`📱 Connected as: ${info.pushname}`);
      console.log(`📞 Phone: ${info.wid.user}`);
      console.log('\n🚀 Agent is now ready to send and receive messages!\n');
    });

    // Disconnected
    this.client.on('disconnected', (reason) => {
      console.log('❌ Client was disconnected:', reason);
      this.isReady = false;
    });

    // Receive messages
    this.client.on('message', async (message) => {
      await this.handleIncomingMessage(message);
    });

    // Message creation (sent messages)
    this.client.on('message_create', async (message) => {
      // Handle sent messages if needed
      if (message.fromMe) {
        console.log(`📤 Sent message to ${message.to}: ${message.body}`);
      }
    });
  }

  // Start the client
  async start() {
    console.log('🚀 Starting WhatsApp Agent...');
    await this.client.initialize();
  }

  // Stop the client
  async stop() {
    console.log('🛑 Stopping WhatsApp Agent...');
    await this.client.destroy();
    this.isReady = false;
  }

  // Wait for client to be ready
  async waitForReady(timeout = 60000) {
    const startTime = Date.now();
    while (!this.isReady) {
      if (Date.now() - startTime > timeout) {
        throw new Error('Timeout waiting for WhatsApp client to be ready');
      }
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    return true;
  }

  // ============================================
  // SEND SMS (TEXT MESSAGES)
  // ============================================
  
  async sendMessage(phoneNumber, message) {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const chat = await this.client.getChatById(chatId);
      const result = await chat.sendMessage(message);
      console.log(`✅ Message sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send message to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // SEND IMAGES
  // ============================================
  
  async sendImage(phoneNumber, imagePath, caption = '') {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const media = MessageMedia.fromFilePath(imagePath);
      const result = await this.client.sendMessage(chatId, media, { caption });
      console.log(`✅ Image sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send image to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // SEND DOCUMENTS
  // ============================================
  
  async sendDocument(phoneNumber, documentPath, caption = '') {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const media = MessageMedia.fromFilePath(documentPath);
      const result = await this.client.sendMessage(chatId, media, { 
        caption,
        sendMediaAsDocument: true 
      });
      console.log(`✅ Document sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send document to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // SEND MEDIA FROM URL
  // ============================================
  
  async sendMediaFromUrl(phoneNumber, url, caption = '', options = {}) {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const media = await MessageMedia.fromUrl(url, options);
      const result = await this.client.sendMessage(chatId, media, { caption });
      console.log(`✅ Media from URL sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send media from URL to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // BROADCAST MESSAGES
  // ============================================
  
  async broadcast(phoneNumbers, message, delay = 1000) {
    const results = [];
    console.log(`📢 Starting broadcast to ${phoneNumbers.length} contacts...`);
    
    for (let i = 0; i < phoneNumbers.length; i++) {
      const phoneNumber = phoneNumbers[i];
      console.log(`📤 Sending to ${phoneNumber} (${i + 1}/${phoneNumbers.length})...`);
      
      const result = await this.sendMessage(phoneNumber, message);
      results.push({
        phoneNumber,
        ...result
      });
      
      // Delay between messages to avoid spam detection
      if (i < phoneNumbers.length - 1) {
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
    
    const successful = results.filter(r => r.success).length;
    console.log(`✅ Broadcast complete: ${successful}/${phoneNumbers.length} successful`);
    
    return {
      total: phoneNumbers.length,
      successful,
      failed: phoneNumbers.length - successful,
      results
    };
  }

  // ============================================
  // BROADCAST IMAGES
  // ============================================
  
  async broadcastImage(phoneNumbers, imagePath, caption = '', delay = 1000) {
    const results = [];
    console.log(`📢 Starting image broadcast to ${phoneNumbers.length} contacts...`);
    
    for (let i = 0; i < phoneNumbers.length; i++) {
      const phoneNumber = phoneNumbers[i];
      console.log(`📤 Sending image to ${phoneNumber} (${i + 1}/${phoneNumbers.length})...`);
      
      const result = await this.sendImage(phoneNumber, imagePath, caption);
      results.push({
        phoneNumber,
        ...result
      });
      
      if (i < phoneNumbers.length - 1) {
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
    
    const successful = results.filter(r => r.success).length;
    console.log(`✅ Image broadcast complete: ${successful}/${phoneNumbers.length} successful`);
    
    return {
      total: phoneNumbers.length,
      successful,
      failed: phoneNumbers.length - successful,
      results
    };
  }

  // ============================================
  // RECEIVE MESSAGES - Add custom handlers
  // ============================================
  
  onMessage(handler) {
    this.messageHandlers.push(handler);
  }

  async handleIncomingMessage(message) {
    // Log incoming message
    const contact = await message.getContact();
    console.log(`📨 Received from ${contact.pushname || contact.number}: ${message.body}`);
    
    // Call all registered handlers
    for (const handler of this.messageHandlers) {
      try {
        await handler(message, contact);
      } catch (error) {
        console.error('Error in message handler:', error);
      }
    }
  }

  // ============================================
  // GET CHATS
  // ============================================
  
  async getChats() {
    try {
      const chats = await this.client.getChats();
      return chats;
    } catch (error) {
      console.error('Failed to get chats:', error.message);
      return [];
    }
  }

  // ============================================
  // GET CONTACTS
  // ============================================
  
  async getContacts() {
    try {
      const contacts = await this.client.getContacts();
      return contacts;
    } catch (error) {
      console.error('Failed to get contacts:', error.message);
      return [];
    }
  }

  // ============================================
  // SEND LOCATION
  // ============================================
  
  async sendLocation(phoneNumber, latitude, longitude, description = '') {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const location = {
        location: {
          lat: latitude,
          lng: longitude,
          description
        }
      };
      const result = await this.client.sendMessage(chatId, location);
      console.log(`✅ Location sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send location to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // SEND CONTACT CARD
  // ============================================
  
  async sendContact(phoneNumber, contactNumber, contactName) {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const contactId = this.formatPhoneNumber(contactNumber);
      const contact = await this.client.getContactById(contactId);
      const result = await this.client.sendMessage(chatId, contact);
      console.log(`✅ Contact card sent to ${phoneNumber}`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send contact to ${phoneNumber}:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // GROUP OPERATIONS
  // ============================================
  
  async createGroup(name, participants) {
    try {
      const participantIds = participants.map(p => this.formatPhoneNumber(p));
      const result = await this.client.createGroup(name, participantIds);
      console.log(`✅ Group created: ${name}`);
      return { success: true, groupId: result.gid._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to create group:`, error.message);
      return { success: false, error: error.message };
    }
  }

  async sendGroupMessage(groupId, message) {
    try {
      const result = await this.client.sendMessage(groupId, message);
      console.log(`✅ Group message sent`);
      return { success: true, messageId: result.id._serialized, data: result };
    } catch (error) {
      console.error(`❌ Failed to send group message:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // STATUS/STORY
  // ============================================
  
  async sendStatus(text) {
    try {
      const result = await this.client.sendMessage('status@broadcast', text);
      console.log(`✅ Status posted`);
      return { success: true, data: result };
    } catch (error) {
      console.error(`❌ Failed to post status:`, error.message);
      return { success: false, error: error.message };
    }
  }

  // ============================================
  // UTILITY FUNCTIONS
  // ============================================
  
  formatPhoneNumber(phoneNumber) {
    // Remove all non-numeric characters
    let cleaned = phoneNumber.replace(/\D/g, '');
    
    // Add country code if not present (assuming India +91, change as needed)
    if (!cleaned.startsWith('91') && cleaned.length === 10) {
      cleaned = '91' + cleaned;
    }
    
    // Format as WhatsApp chat ID
    return `${cleaned}@c.us`;
  }

  async getProfilePicUrl(phoneNumber) {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const url = await this.client.getProfilePicUrl(chatId);
      return url;
    } catch (error) {
      console.error(`Failed to get profile picture:`, error.message);
      return null;
    }
  }

  async isRegisteredUser(phoneNumber) {
    try {
      const chatId = this.formatPhoneNumber(phoneNumber);
      const isRegistered = await this.client.isRegisteredUser(chatId);
      return isRegistered;
    } catch (error) {
      console.error(`Failed to check if user is registered:`, error.message);
      return false;
    }
  }

  getState() {
    return this.client.getState();
  }

  async getInfo() {
    return this.client.info;
  }
}

export default WhatsAppAgent;

File: agents/whatsapp_agent.py
"""
WhatsApp Agent - Bridge between Python agents and Node.js WhatsApp service
Handles sending messages, images, documents via WhatsApp
"""

import requests
import json
from typing import Dict, Any, Optional, List
from pathlib import Path
import os

class WhatsAppAgent:
    """
    WhatsApp Agent for sending messages and media via WhatsApp Web
    Communicates with Node.js WhatsApp service
    """
    
    def __init__(self, api_url: str = "http://localhost:3000"):
        self.api_url = api_url
        self.name = "WhatsApp Agent"
        self.description = "Send messages, images, documents, and more via WhatsApp"
    
    def _check_service_health(self) -> bool:
        """Check if WhatsApp service is running and ready"""
        try:
            response = requests.get(f"{self.api_url}/health", timeout=2)
            data = response.json()
            return data.get('ready', False)
        except:
            return False
    
    def _format_phone_number(self, phone: str) -> str:
        """Format phone number - remove '+' and ensure proper format"""
        # Remove all non-numeric characters (including +)
        phone = ''.join(filter(str.isdigit, phone))
        
        # Validate: should be 12 digits (country code + 10 digits)
        # If only 10 digits, user needs to add country code
        if len(phone) == 10:
            return None  # Return None to indicate error - need country code
        elif len(phone) == 12 and phone.startswith('91'):
            return phone  # Valid Indian number
        elif len(phone) >= 11:
            return phone  # Assume valid international number
        else:
            return None  # Invalid format
    
    def send_message(self, phone_number: str, message: str) -> Dict[str, Any]:
        """
        Send a text message via WhatsApp
        
        Args:
            phone_number: Phone number (with or without country code)
            message: Text message to send
            
        Returns:
            Response with success status and message ID
        """
        try:
            if not self._check_service_health():
                return {
                    'status': 'error',
                    'message': 'WhatsApp service is not ready. Please start the WhatsApp agent first.',
                    'service_url': self.api_url
                }
            
            phone_number = self._format_phone_number(phone_number)
            
            # Check if phone number formatting failed
            if phone_number is None:
                return {
                    'status': 'error',
                    'message': 'Invalid phone number format. Please provide number with country code (e.g., 919865324172 for India)',
                }
            
            response = requests.post(
                f"{self.api_url}/send/message",
                json={
                    'phoneNumber': phone_number,
                    'message': message
                },
                timeout=30
            )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    'status': 'success',
                    'message': f'Message sent to +{phone_number}',
                    'data': result
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Failed to send message: {response.text}'
                }
                
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Error sending WhatsApp message: {str(e)}'
            }
    
    def send_image(self, phone_number: str, image_path: str, caption: str = '') -> Dict[str, Any]:
        """
        Send an image via WhatsApp
        
        Args:
            phone_number: Phone number
            image_path: Path to image file
            caption: Optional caption for the image
            
        Returns:
            Response with success status
        """
        try:
            if not self._check_service_health():
                return {
                    'status': 'error',
                    'message': 'WhatsApp service is not ready'
                }
            
            if not os.path.exists(image_path):
                return {
                    'status': 'error',
                    'message': f'Image file not found: {image_path}'
                }
            
            phone_number = self._format_phone_number(phone_number)
            
            with open(image_path, 'rb') as f:
                files = {'image': f}
                data = {
                    'phoneNumber': phone_number,
                    'caption': caption
                }
                
                response = requests.post(
                    f"{self.api_url}/send/image",
                    files=files,
                    data=data,
                    timeout=60
                )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    'status': 'success',
                    'message': f'Image sent to +{phone_number}',
                    'data': result
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Failed to send image: {response.text}'
                }
                
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Error sending image: {str(e)}'
            }
    
    def send_document(self, phone_number: str, file_path: str, caption: str = '') -> Dict[str, Any]:
        """
        Send a document via WhatsApp
        
        Args:
            phone_number: Phone number
            file_path: Path to document file
            caption: Optional caption
            
        Returns:
            Response with success status
        """
        try:
            if not self._check_service_health():
                return {
                    'status': 'error',
                    'message': 'WhatsApp service is not ready'
                }
            
            if not os.path.exists(file_path):
                return {
                    'status': 'error',
                    'message': f'File not found: {file_path}'
                }
            
            phone_number = self._format_phone_number(phone_number)
            
            with open(file_path, 'rb') as f:
                files = {'document': f}
                data = {
                    'phoneNumber': phone_number,
                    'caption': caption
                }
                
                response = requests.post(
                    f"{self.api_url}/send/document",
                    files=files,
                    data=data,
                    timeout=60
                )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    'status': 'success',
                    'message': f'Document sent to +{phone_number}',
                    'data': result
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Failed to send document: {response.text}'
                }
                
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Error sending document: {str(e)}'
            }
    
    def broadcast_message(self, phone_numbers: List[str], message: str) -> Dict[str, Any]:
        """
        Broadcast message to multiple contacts
        
        Args:
            phone_numbers: List of phone numbers
            message: Message to broadcast
            
        Returns:
            Response with success status and results
        """
        try:
            if not self._check_service_health():
                return {
                    'status': 'error',
                    'message': 'WhatsApp service is not ready'
                }
            
            formatted_numbers = [self._format_phone_number(num) for num in phone_numbers]
            
            response = requests.post(
                f"{self.api_url}/send/broadcast",
                json={
                    'phoneNumbers': formatted_numbers,
                    'message': message
                },
                timeout=60
            )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    'status': 'success',
                    'message': f'Broadcast sent to {len(phone_numbers)} contacts',
                    'data': result
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Failed to broadcast: {response.text}'
                }
                
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Error broadcasting message: {str(e)}'
            }
    
    def verify_number(self, phone_number: str) -> Dict[str, Any]:
        """Check if a number is registered on WhatsApp"""
        try:
            if not self._check_service_health():
                return {
                    'status': 'error',
                    'message': 'WhatsApp service is not ready'
                }
            
            phone_number = self._format_phone_number(phone_number)
            
            response = requests.post(
                f"{self.api_url}/verify/number",
                json={'phoneNumber': phone_number},
                timeout=10
            )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    'status': 'success',
                    'data': result
                }
            else:
                return {
                    'status': 'error',
                    'message': f'Failed to verify: {response.text}'
                }
                
        except Exception as e:
            return {
                'status': 'error',
                'message': f'Error verifying number: {str(e)}'
            }
    
    def process(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """
        Process WhatsApp request
        
        Supported operations:
        - send_message: Send text message
        - send_image: Send image with optional caption
        - send_document: Send document/file
        - broadcast: Send to multiple contacts
        - verify: Check if number is on WhatsApp
        """
        operation = request.get('operation', 'send_message')
        
        if operation == 'send_message':
            phone = request.get('phone_number')
            message = request.get('message')
            
            if not phone or not message:
                return {
                    'status': 'error',
                    'message': 'phone_number and message are required'
                }
            
            return self.send_message(phone, message)
        
        elif operation == 'send_image':
            phone = request.get('phone_number')
            image_path = request.get('image_path')
            caption = request.get('caption', '')
            
            if not phone or not image_path:
                return {
                    'status': 'error',
                    'message': 'phone_number and image_path are required'
                }
            
            return self.send_image(phone, image_path, caption)
        
        elif operation == 'send_document':
            phone = request.get('phone_number')
            file_path = request.get('file_path')
            caption = request.get('caption', '')
            
            if not phone or not file_path:
                return {
                    'status': 'error',
                    'message': 'phone_number and file_path are required'
                }
            
            return self.send_document(phone, file_path, caption)
        
        elif operation == 'broadcast':
            phones = request.get('phone_numbers', [])
            message = request.get('message')
            
            if not phones or not message:
                return {
                    'status': 'error',
                    'message': 'phone_numbers and message are required'
                }
            
            return self.broadcast_message(phones, message)
        
        elif operation == 'verify':
            phone = request.get('phone_number')
            
            if not phone:
                return {
                    'status': 'error',
                    'message': 'phone_number is required'
                }
            
            return self.verify_number(phone)
        
        else:
            return {
                'status': 'error',
                'message': f'Unknown operation: {operation}'
            }


# For testing
if __name__ == "__main__":
    agent = WhatsAppAgent()
    
    # Test health check
    print("Testing WhatsApp service connection...")
    if agent._check_service_health():
        print("✅ WhatsApp service is ready!")
    else:
        print("❌ WhatsApp service is not ready. Please start it first.")
        print(f"   Run: cd Whatsapp-Agent && npm start")

